### PR TITLE
Add `#[derive(TypedUuidKind)]`

### PR DESCRIPTION
**Pull request by hawkw**
This branch addss a simple implementation of a derive macro for the `TypedUuidKind` trait. By default, the type name is used as the tag, but it can be overridden by adding a `#[typed_uuid_tag("tag")]` attribute to the type deriving `TypedUuidKind`.

### CODE REVIEW

File: crates/integration-tests/Cargo.toml [hunk 1]
```diff
@@ -29,3 +31,6 @@ internal-schemars08-tests = [
     "dep:syn",
     "dep:typify",
 ]
+internal-derive-tests = [
+    "newtype-uuid/derive",
+]
```
Comment (sunshowers) [inline crates/integration-tests/Cargo.toml:36]: Is this still required?

**Review by sunshowers — COMMENTED** [review]
Thanks!

File: crates/integration-tests/Cargo.toml (hunk 1 above)
Comment (hawkw) [reply crates/integration-tests/Cargo.toml:36]: Ah, no, it is not. Thanks.
```suggestion
```

**Comment by hawkw** [comment]
@sunshowers anything else left to address here?

**Comment by sunshowers** [comment]
Thanks!

**Comment by sunshowers** [comment]
Had a chat with Eliza, and here's the conclusion we came to:

* By default, the kind is the name of the type, which will be in TitleCase per Rust rules.
* This seems generally correct -- for e.g. serde, the default name generated for an enum variant is in TitleCase
* But almost all of the time, we'll want the tag to be in snake or kebab case
* This means that the tag will have to be manually specified almost all of the time.
* This increases verbosity.

For example, let's say we want to define these three kinds:

* `DownstairsKind`
* `DownstairsRegionKind`
* `LoopbackAddressKind`

With this proc macro, it would look something like:

```rust
#[derive(TypedUuidKind)]
#[typed_uuid_tag("downstairs_kind")]
pub enum DownstairsKind {}

#[derive(TypedUuidKind)]
#[typed_uuid_tag("downstairs_region_kind")]
pub enum DownstairsRegionKind {}

#[derive(TypedUuidKind)]
#[typed_uuid_tag("loopback_address_kind")]
pub enum LoopbackAddressKind {}
```

Compare this to the suggestion to use a macro here: https://docs.rs/newtype-uuid/1.0.1/newtype_uuid/trait.TypedUuidKind.html#notes. Adding a new kind becomes just 1 line of code, and amortizes very well with a larger number of kinds.

```rust
impl_typed_uuid_kind! {
    DownstairsKind => "downstairs",
    DownstairsRegionKind => "downstairs_region",
    LoopbackAddressKind => "loopback_address",
}
```

Users could write a declarative macro that wraps the proc-macro, but at that point why not just write out the full declarative macro?

The proc macro just doesn't seem to win us much relative to the complexity it introduces.

In the future we may want this if `TypedUuidKind` becomes more complex, so let's keep the branch around -- but let's not land this in main.

Thanks again @hawkw for putting in all this effort!

**Pull request closed without merging**